pallets/nft-manager/src/lib.rs

```diff
@@ -789,65 +791,34 @@ pub mod pallet {
             );
 
             if onchain_version < 4 {
-                use frame_support::storage::unhashed;
-
-                // Owned Nfts cleanup
-                {
-                    let owned_nfts_prefix = storage::storage_prefix(b"NftManager", b"OwnedNfts");
-                    let mut key = vec![0u8; 32];
-                    key[0..32].copy_from_slice(&owned_nfts_prefix);
-                    let res = unhashed::clear_prefix(&key[0..32], None, None);
-
-                    log::info!(
-                    "✅ Cleared '{}' backend values from 'NftManager::OwnedNfts' storage prefix",
-                        res.backend
-                    );
-                    log::info!(
-                        "✅ Cleared '{}' entries from 'NftManager::OwnedNfts' storage prefix",
-                        res.unique
-                    );
-                    if res.maybe_cursor.is_some() {
-                        log::error!(
-                            "Storage prefix 'NftManager::OwnedNfts' is not completely cleared."
-                        );
-                    }
-                }
-                // Version item cleanup
-                {
-                    let storage_version_prefix =
-                        storage::storage_prefix(b"NftManager", b"StorageVersion");
-                    let mut key = vec![0u8; 32];
-                    key[0..32].copy_from_slice(&storage_version_prefix);
-                    let res = unhashed::clear_prefix(&key[0..32], None, None);
-
-                    log::info!(
-                    "✅ Cleared '{}' backend values from 'NftManager::StorageVersion' storage prefix",
-                        res.backend
-                    );
-                    log::info!(
-                        "✅ Cleared '{}' entries from 'NftManager::StorageVersion' storage prefix",
-                        res.unique
-                    );
-                    if res.maybe_cursor.is_some() {
-                        log::error!("Storage prefix 'StorageVersion' is not completely cleared.");
-                    }
-                }
-                crate::STORAGE_VERSION.put::<Pallet<T>>();
+                let mut migration_weight: Weight = Weight::from_ref_time(0);
+
+                migration_weight += migrations::partial_owned_nfts_entries_drop::<T>();
+                migration_weight += migrations::safe_storage_info_drop::<T>();
+
                 log::info!(
                     "Dropped old enum-based versioning schema for nft-manager. New version:{:?}",
                     Pallet::<T>::on_chain_storage_version()
                 );
 
-                return migrations::migrate_to_bounded_nft::<T>()
+                return migration_weight
             }
             return Weight::from_ref_time(0)
         }
+
+        fn on_initialize(_n: T::BlockNumber) -> Weight {
```

Comment:
what is this hook? how is this different from `on_runtime_upgrade`

Reply:
This is a hook that is called on the initialisation of each block, allowing you to perform actions on system-reserved space for each block

Reply:
I suspect that `on_runtime_upgrade` will run once, when an upgrade is made, but `on_initialize` will run on any new block?